pallets/staking/src/lib.rs

```diff
@@ -0,0 +1,923 @@
+}
+
+impl<T: Config> Pallet<T> {
+	/// Account id holding rewards to pay.
+	pub fn pot_account_id() -> T::AccountId {
+		T::PalletId::get().into_account_truncating()
+	}
+
+	fn ensure_stakable_balance(
+		who: &T::AccountId,
+		stake: Balance,
+		position: Option<&Position<T::BlockNumber>>,
+	) -> Result<(), DispatchError> {
+		let free_balance = T::Currency::free_balance(T::HdxAssetId::get(), who);
+		let staked = if let Some(p) = position { p.stake } else { Zero::zero() };
+		let vested = T::Vesting::locked(who.clone());
+
+		let stakable = free_balance
+			.checked_sub(vested)
+			.ok_or(Error::<T>::Arithmetic)?
+			.checked_sub(staked)
+			.ok_or(Error::<T>::Arithmetic)?;
+
+		ensure!(stakable >= stake, Error::<T>::InsufficientBalance);
+
+		Ok(())
+	}
+
+	/// Returns staking `PositionItemId` for `who` or `None`.
+	pub fn get_user_position_id(who: &T::AccountId) -> Result<Option<T::PositionItemId>, DispatchError> {
+		let mut user_position_ids = T::NFTHandler::owned_in_collection(&T::NFTCollectionId::get(), who);
+
+		let position_id = user_position_ids.next();
+		if position_id.is_some() {
+			ensure!(
+				user_position_ids.next().is_none(),
+				Error::<T>::InconsistentState(InconsistentStateError::TooManyPostions)
+			);
+
+			return Ok(position_id);
+		}
+
+		Ok(None)
+	}
+
+	fn is_owner(who: &T::AccountId, id: T::PositionItemId) -> bool {
+		if let Some(owner) =
+			<T as pallet::Config>::NFTHandler::owner(&<T as pallet::Config>::NFTCollectionId::get(), &id)
+		{
+			owner == *who
+		} else {
+			false
+		}
+	}
+
+	fn create_position_and_mint_nft(
+		who: &T::AccountId,
+		staked_amount: Balance,
+		accumulated_reward_per_stake: FixedU128,
+	) -> Result<T::PositionItemId, DispatchError> {
+		let position_id = Self::get_next_position_id()?;
+		Positions::<T>::insert(
+			position_id,
+			Position::new(
+				staked_amount,
+				accumulated_reward_per_stake,
+				T::BlockNumberProvider::current_block_number(),
+			),
+		);
+
+		T::NFTHandler::mint_into(&T::NFTCollectionId::get(), &position_id, who)?;
+
+		Ok(position_id)
+	}
+
+	fn get_next_position_id() -> Result<T::PositionItemId, DispatchError> {
+		<NextPositionId<T>>::try_mutate(|current_value| -> Result<T::PositionItemId, DispatchError> {
+			let next_id = *current_value;
+
+			*current_value = current_value
+				.checked_add(&T::PositionItemId::one())
+				.ok_or(Error::<T>::Arithmetic)?;
+
+			Ok(next_id)
+		})
+	}
+
+	/// This function "distributes" pending rewards if possible and updates `StakingData`
+	fn update_rewards(staking: &mut StakingData) -> Result<(), DispatchError> {
+		if staking.total_stake.is_zero() {
+			return Ok(());
+		}
+
+		let pending_rewards = T::Currency::free_balance(T::HdxAssetId::get(), &Self::pot_account_id())
+			.checked_sub(staking.pot_reserved_balance)
+			.defensive_ok_or::<Error<T>>(InconsistentStateError::NegativePendingRewards.into())?;
+
+		if pending_rewards.is_zero() {
+			return Ok(());
```

Comment:
It's tested here: https://github.com/galacticcouncil/HydraDX-node/blob/71814fa3d2f564c2918fe6cd18b797d73520bb59/pallets/staking/src/tests/stake.rs#L148****

Reply:
The issue is that if I remove these three lines of code (756-758), this test still pass. So it is a sort of false negative.

I wonder if we could meaningfully fix this test somehow, possibly tweaking the test setup. 

Or if it would lead to an unrealistic test scenario, then leave it.

Reply:
(756-758) is there only to prevent unnecessary calculation. Calculations still works with `0` but  they are just not necessary, that's why this tests still passes.
I don't think this can be tested because calculation returns `0` for zero pending rewards so nothing would change if you remove if's from this function.
I added test for event that will fail if you remove all the if's with early return.
https://github.com/galacticcouncil/HydraDX-node/blob/06d8ebd0dacb7b9edf2e0c5bd92dd650fc411cdf/pallets/staking/src/tests/tests.rs#L476

Reply:
ah okay, I did not know that it was a sort of performance optimization.

fine then, thanks for clarifying!